File: java/759. Employee Free Time.java
/*
// Definition for an Interval.
class Interval {
    public int start;
    public int end;

    public Interval() {}

    public Interval(int _start, int _end) {
        start = _start;
        end = _end;
    }
};
*/
class Solution {
    public List<Interval> employeeFreeTime(List<List<Interval>> schedule) {
        List<Interval> total = new ArrayList<>();
        for (List<Interval> list : schedule) {
            total.addAll(list);
        }

        total.sort((i1, i2) -> {
            if (i1.start != i2.start)
                return i1.start - i2.start;
            else {
                return i1.end - i2.end;
            }
        });
        // ??
        if (total.size() <= 1) {
            return total;
        }
        Interval pre = total.get(0);
        List<Interval> freeTime = new ArrayList<Interval>();
        for (int i = 1; i < total.size(); i++) {
            Interval cur = total.get(i);
            if (cur.start > pre.end) {
                freeTime.add(new Interval(pre.end, cur.start));
                pre = cur;
            } else {
                if (cur.end > pre.end)
                    pre.end = cur.end;
            }
        }
        return freeTime;

    }
}

File: js/leetcode/js/759. Employee Free Time.js
/*
 * @lc app=leetcode.cn id=759 lang=javascript
 *
 * [759] Employee Free Time
 */

// @lc code=start
/**
 * // Definition for an Interval.
 * function Interval(start, end) {
 * this.start = start;
 * this.end = end;
 * };
 * /

/**
 * @param {Interval[][]} schedule
 * @return {Interval[]}
 */
var employeeFreeTime = function (schedule) {
    schedule = schedule.reduce((pre, cur) => pre.concat(cur))

    const merged = [];
    schedule.sort((a, b) => {
        if (a.start !== b.start) {
            return a.start - b.start
        } else {
            return a.end - b.end;
        }
    })
    const freeIntervals = [];

    // free time
    for (let i = 0; i < schedule.length; i++) {
        const cur = schedule[i]
        // last.end > cur.start

        if (!merged.length) {
            merged.push(cur)
        } else {
            const last = merged[merged.length - 1]
            if (last.end >= cur.start) {
                last.end = Math.max(cur.end, last.end)
            } else {
                merged.push(cur)
                freeIntervals.push(new Interval(last.end, cur.start))
            }
        }
    }
    return freeIntervals;

};
schedule =
    [[[1, 2], [5, 6]], [[1, 3]], [[4, 10]]]

// @lc code=end

